custom/enikshay/two_b_release_1/management/commands/enikshay_2b_case_properties.py

```diff
@@ -0,0 +1,496 @@
+            "real" if commit else "fake", domain, datetime.datetime.utcnow()
+        ))
+        dto = SQLLocation.objects.get(
+            domain=domain, location_id=dto_id, location_type__code='dto')
+        num_descendants = dto.get_descendants(include_self=True).count()
+        confirm("Do you want to migrate the DTO '{}', which has {} descendants?"
+                .format(dto.get_path_display(), num_descendants))
+        migrator = ENikshay2BMigrator(domain, dto, commit)
+        migrator.migrate()
+        logger.info("Migrated {} person cases".format(migrator.total_persons))
+        logger.info("Migrated {} occurrence cases".format(migrator.total_occurrences))
+        logger.info("Migrated {} episode cases".format(migrator.total_episodes))
+        logger.info("Migrated {} test cases".format(migrator.total_tests))
+        logger.info("Migrated {} referral cases".format(migrator.total_referrals))
+        logger.info("Migrated {} trail cases".format(migrator.total_trails))
+        logger.info("Created {} secondary_owner cases".format(migrator.total_secondary_owners))
+        logger.info("Closed {} drtb_hiv cases".format(migrator.total_drtb_hiv))
+
+
+class ENikshay2BMigrator(object):
+    def __init__(self, domain, dto, commit):
+        self.domain = domain
+        self.dto = dto
+        self.commit = commit
+        self.accessor = CaseAccessors(self.domain)
+        self.factory = CaseFactory(self.domain)
+
+        self.total_persons = 0
+        self.total_occurrences = 0
+        self.total_episodes = 0
+        self.total_tests = 0
+        self.total_referrals = 0
+        self.total_trails = 0
+        self.total_secondary_owners = 0
+        self.total_drtb_hiv = 0
+
+    @property
+    @memoized
+    def locations(self):
+        return {loc.location_id: loc for loc in
+                self.dto.get_descendants(include_self=True).prefetch_related('location_type')}
+
+    @property
+    @memoized
+    def location_ids_by_pk(self):
+        return {loc.pk: loc.location_id for loc in self.locations.values()}
+
+    def get_ancestors_by_type(self, location):
+        """Get all direct ancestors found in self.locations"""
+        ancestors_by_type = {location.location_type.code: location}
+        loc = location
+        while loc.parent_id and loc.parent_id in self.location_ids_by_pk:
+            parent = self.locations[self.location_ids_by_pk[loc.parent_id]]
+            ancestors_by_type[parent.location_type.code] = parent
+            loc = parent
+        return ancestors_by_type
+
+    def migrate(self):
+        person_ids = self.get_relevant_person_case_ids()
+        persons = self.get_relevant_person_case_sets(person_ids)
+        for person in with_progress_bar(persons, len(person_ids)):
+            self.migrate_person_case_set(person)
+
+    def get_relevant_person_case_ids(self):
+        location_owners = self.dto.get_descendants(include_self=True).location_ids()
```

Comment:
I definitely could do that - made those two changes too far apart to notice, thanks!